### ISSUE TITLE
zoom on y

### ISSUE THREAD

**Issue by itsjimbo** [rrag/react-stockcharts#101]
I took a stab at the zoom on y-axis, and writing plunkr with react is a bit challenging, but this small component works for my use-case scenario.   Any comments/suggestions welcome..

Cheers!

## Plunkr
http://embed.plnkr.co/1pkiYEumHka5H4d7MeNk/

### Added the following CSS
```
.draggable {
  stroke-width: 4;
  cursor: move;
}
.hover_group_drag:hover rect {
  fill-opacity: .3;
  stroke-width: 4;
  opacity: .3;
  fill: #000;
}
```

### Screenshot

![image](https://cloud.githubusercontent.com/assets/11524813/16759669/ed031832-47cc-11e6-8d6c-a5e8c2ea79ab.png)

**Comment by rrag** [comment]
This looks great.

After zooming on yAxis if you pan the main chart, I think you would want to maintain the same domain rather than keep the same magnification factor.

Also the magnification does not go above 1, is that intentional?

Here is a modified version http://plnkr.co/edit/9Ix2AIfH7Aa6GN1TCNU7?p=preview

One additional thing I would like to do that the zoom does not get interrupted because of mouse leaving the magnification area, similar to how the pan works.

**Comment by rrag** [comment]
Also once you zoom on y, it would be desirable if the pan can work on both x and y, instead of the default x only pan, this will have to be done with updates to the main library itself.

**Comment by itsjimbo** [comment]
@rrag - the original intent of the -1/+1 magnification was to prohibit values too far out of range, but this also can be done elsewhere, eg (financial data does not need to go below zero, but sensor data should).

I see your changes and I guess upon first inspection
```
var yExtents = newYDomain || (d => [d.high, d.low]);
```
This sort of 'locks' the domain now, so there is no autoscaling on panning, my original thoughts were to add an offset to the high and low function, which does not fix the domain, and allows for auto scaling on panning..
```
var yExtents = (d => [d.high +offset, d.low- offset]);
```

**Comment by rrag** [comment]
> the original intent of the -1/+1 magnification was to prohibit values too far out of range

I agree that will be important too. will need some changes to limit the amount of magnification you can do

> This sort of 'locks' the domain now

yeah that is the idea, there could be a reset button somewhere else on the page which will set the `newYDomain` to `null`

tradingview has the same implementation I think

I like how this much is possible without changing anything in the library at all, and the fact you like 

```
var yExtents = (d => [d.high +offset, d.low- offset]);
```
while I prefer
```
var yExtents = newYDomain || (d => [d.high, d.low]);
```
means that this kind of setting if left in userland is likely the right solution

**Comment by itsjimbo** [comment]
Added clear magnification based on your example...

http://run.plnkr.co/plunks/2NQPx1UZW3COxkC2KUyM/

**Comment by itsjimbo** [comment]
also just added different high/low deltas based on the user position, eg, a spot to magnify...
```
    var deltaHigh = (high - yValue) / 30;
    var deltaLow = (yValue - low) / 30; 
```
http://plnkr.co/edit/2NQPx1UZW3COxkC2KUyM?p=preview


but i think my calculation needs to also involve the updated magnitude as it is changing
```
yScale.invert( e.clientY)
```
you can see this behavior if you mousedown and drag up and down, the zoom moves farther away

**Comment by rrag** [comment]
I like this behavior. Now we need to figure out a way to pan on y, when zoom on y is done (`newYDomain` is not null)

**Comment by WaiSiuKei** [comment]
Hi @itsjimbo , you have done a great job!

I am implementing a simailar feature in my repo but doing a little difference. I am using the center of yAxis as a base point for magnification, that will be a little easier for users to align the series to the middle.

[https://waisiukei.github.io/nila/](https://waisiukei.github.io/nila/)
You can pan on both x & y by dragging the chart area, zoom on x by wheel, and zoom on y by dragging the yAxis.

**Comment by rrag** [comment]
Thanks @WaiSiuKei  for the idea of magnification based on center on y. Your library looks very promising, especially since it is framework agnostic.

I have used your idea of zoom based on center y for changes in `0.6`

Here is an work in progress version

http://rrag.github.io/react-stockcharts-next/documentation.html#/edge_coordinate

Once you zoom on y, the pan becomes 2 dimensional. Change to svg and back to canvas, and the yAxis is reset.
you could also zoom on x by dragging the x axis.

**Comment by rrag** [comment]
merged to master and `0.6.0-beta.2` published to npm
